src/crystal/system/unix/urandom.cr

```diff
@@ -1,35 +1,51 @@
-{% skip_file unless flag?(:unix) && !flag?(:openbsd) && !flag?(:linux) %}
+{% skip_file unless flag?(:unix) && !flag?(:openbsd) %}
 
-module Crystal::System::Random
+require "c/fcntl"
+require "c/unistd"
+require "c/sys/stat"
+
+# :nodoc:
+module Crystal::System::Urandom
   @@initialized = false
-  @@urandom : ::File?
+  @@urandom : LibC::Int?
 
   private def self.init
     @@initialized = true
 
-    urandom = ::File.open("/dev/urandom", "r")
-    return unless urandom.stat.chardev?
+    fd = LibC.open("/dev/urandom", LibC::O_RDONLY)
+    raise Errno.new("open") if fd == -1
 
-    urandom.close_on_exec = true
-    urandom.sync = true # don't buffer bytes
-    @@urandom = urandom
-  end
+    raise Errno.new("fstat") if LibC.fstat(fd, out stats) == -1
+    unless (stats.st_mode & LibC::S_IFMT) == LibC::S_IFCHR
+      raise Errno.new("close") if LibC.close(fd) == -1
+      return
+    end
 
-  def self.random_bytes(buf : Bytes) : Nil
-    init unless @@initialized
+    flags = LibC.fcntl(fd, LibC::F_GETFD, 0)
+    raise Errno.new("fcntl") if flags == -1
 
-    if urandom = @@urandom
-      urandom.read_fully(buf)
-    else
-      raise "Failed to access secure source to generate random bytes!"
-    end
+    flags |= LibC::O_NONBLOCK | LibC::FD_CLOEXEC
+    raise Errno.new("fcntl") if LibC.fcntl(fd, LibC::F_SETFD, flags) == -1
+
+    @@urandom = fd
   end
 
-  def self.next_u : UInt8
+  def self.random_bytes(buf : Bytes) : Nil
     init unless @@initialized
 
-    if urandom = @@urandom
-      urandom.read_bytes(UInt8)
+    if fd = @@urandom
+      while buf.size > 0
+        read_bytes = LibC.read(fd, buf, buf.size)
+        if read_bytes < 0
+          if Errno.value == Errno::EINTR || Errno.value == Errno::EAGAIN || Errno.value == Errno::EWOULDBLOCK
+            Fiber.yield
```

Comment:
I highly doubt that urandom would ever block, and even if it did this implementation is far from the correct one using libevent or epoll. I'd recommend just not setting nonblock at all.

Reply:
Maybe filling a large slice will, but the call may be interrupted anyway, so we should rescue and yield to the scheduler (there is probably a signal handler pending). It's not ideal, but it does the job —and there are no easy primitives to access the libevent base, or create a simple fd event :cry: